In [2]:
import pandas as pd
from matplotlib import pyplot as plt
from sklearn.preprocessing import StandardScaler
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, LogisticRegression
df = pd.read_csv("diabetes.csv")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 253680 entries, 0 to 253679
Data columns (total 22 columns):
 #   Column                 Non-Null Count   Dtype
---  ------                 --------------   -----
 0   Diabetes               253680 non-null  int64
 1   HighBP                 253680 non-null  int64
 2   HighChol               253680 non-null  int64
 3   BMI                    253680 non-null  int64
 4   Smoker                 253680 non-null  int64
 5   Stroke                 253680 non-null  int64
 6   Myocardial             253680 non-null  int64
 7   PhysActivity           253680 non-null  int64
 8   Fruit                  253680 non-null  int64
 9   Vegetables             253680 non-null  int64
 10  HeavyDrinker           253680 non-null  int64
 11  HasHealthcare          253680 non-null  int64
 12  NotAbleToAffordDoctor  253680 non-null  int64
 13  GeneralHealth          253680 non-null  int64
 14  MentalHealth           253680 non-null  int64
 15  PhysicalHealth   

In [75]:
df.head()

Diabetes  HighBP  HighChol  BMI  Smoker  Stroke  Myocardial  PhysActivity  \
0         0       1         1   40       1       0           0             0   
1         0       0         0   25       1       0           0             1   
2         0       1         1   28       0       0           0             0   
3         0       1         0   27       0       0           0             1   
4         0       1         1   24       0       0           0             1   

   Fruit  Vegetables  HeavyDrinker  HasHealthcare  NotAbleToAffordDoctor  \
0      0           1             0              1                      0   
1      0           0             0              0                      1   
2      1           0             0              1                      1   
3      1           1             0              1                      0   
4      1           1             0              1                      0   

   GeneralHealth  MentalHealth  PhysicalHealth  HardToClimbStairs  \
0              5            18              15                  1   
1              3             0               0                  0   
2              5            30              30                  1   
3              2             0               0                  0   
4              2             3               0                  0   

   BiologicalSex  AgeBracket  EducationBracket  IncomeBracket  Zodiac  
0              1           9                 4              3      10  
1              1           7                 6              1      11  
2              1           9                 4              8       2  
3              1          11                 3              6      11  
4              1          11                 5              4       8

In [76]:
if df.isnull().any().any():
    print('The dataframe has missing values')
else:
    print('The dataframe does not have missing values')

The dataframe does not have missing values


In [77]:
from sklearn import metrics
from sklearn import model_selection
def logistic_regersion(X, y, random_state, name):
    x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=random_state)
    model = LogisticRegression(class_weight='balanced').fit(x_train, y_train)
    y_pred = model.predict(x_test)
    accuracy = metrics.accuracy_score(y_test, y_pred)
    precision = metrics.precision_score(y_test, y_pred)
    recall = metrics.recall_score(y_test, y_pred)
    f1 = metrics.f1_score(y_test, y_pred)
    if name !='':
        print('After dropping ', name)
    print(f'accuracy = {100 * accuracy:0.4f}%')
    print(f'precision = {100 * precision:0.4f}%')
    print(f'recall = {recall:0.4f}')
    print(f'f1 = {f1:0.4f}')
    y_prob = model.predict_proba(x_test)[:, 1]
    fpr, tpr, thresholds = metrics.roc_curve(y_test, y_prob)
    auc_roc = metrics.auc(fpr, tpr)
    print('*AUC is ', auc_roc)
    print('') 
    return model

In [3]:
predictors = ['HighBP', 'HighChol', 'BMI', 'Smoker','Stroke','Myocardial','PhysActivity','Fruit','Vegetables','HeavyDrinker','HasHealthcare','NotAbleToAffordDoctor','GeneralHealth','MentalHealth','PhysicalHealth','HardToClimbStairs', 'BiologicalSex', 'AgeBracket', 'EducationBracket', 'IncomeBracket', 'Zodiac']
X = df[predictors]
y = df['Diabetes']
scaled = StandardScaler().fit_transform(X)
scaled = pd.DataFrame(scaled, columns=X.columns)
(len(X), len(y))


(253680, 253680)

In [4]:
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
(len(x_train), len(y_train))

(202944, 202944)

In [80]:

model1 = logistic_regersion(scaled, y, 0, '')
for i in range(len(predictors)):
    temp = predictors[:]
    temp.pop(i)
    model2 = logistic_regersion(df[temp], y, 0,predictors[i])

accuracy = 73.1690%
precision = 31.7336%
recall = 0.7712
f1 = 0.4496
*AUC is  0.8251768490461993



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  HighBP
accuracy = 72.0140%
precision = 30.4870%
recall = 0.7570
f1 = 0.4347
*AUC is  0.8127932618739113



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  HighChol
accuracy = 72.0297%
precision = 30.5506%
recall = 0.7602
f1 = 0.4359
*AUC is  0.812742409384603



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  BMI
accuracy = 71.9765%
precision = 30.4852%
recall = 0.7590
f1 = 0.4350
*AUC is  0.8097925762948638



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  Smoker
accuracy = 72.5481%
precision = 31.1416%
recall = 0.7691
f1 = 0.4433
*AUC is  0.8184949823371992



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  Stroke
accuracy = 72.0928%
precision = 30.7875%
recall = 0.7720
f1 = 0.4402
*AUC is  0.8186495502634903



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  Myocardial
accuracy = 72.6190%
precision = 31.3098%
recall = 0.7760
f1 = 0.4462
*AUC is  0.8218826269872492



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  PhysActivity
accuracy = 72.3707%
precision = 31.0654%
recall = 0.7744
f1 = 0.4434
*AUC is  0.82112631867629



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  Fruit
accuracy = 72.2229%
precision = 30.9553%
recall = 0.7756
f1 = 0.4425
*AUC is  0.8208960734011659



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  Vegetables
accuracy = 72.5639%
precision = 31.2399%
recall = 0.7746
f1 = 0.4452
*AUC is  0.8214806659460134



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  HeavyDrinker
accuracy = 72.8299%
precision = 31.3493%
recall = 0.7662
f1 = 0.4449
*AUC is  0.8202993559762668



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  HasHealthcare
accuracy = 72.9561%
precision = 31.5602%
recall = 0.7726
f1 = 0.4481
*AUC is  0.8229931248536861



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  NotAbleToAffordDoctor
accuracy = 72.6210%
precision = 31.2296%
recall = 0.7706
f1 = 0.4445
*AUC is  0.8194928085091774



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  GeneralHealth
accuracy = 71.6355%
precision = 30.0489%
recall = 0.7498
f1 = 0.4290
*AUC is  0.8060630740689025



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  MentalHealth
accuracy = 72.4062%
precision = 31.0796%
recall = 0.7733
f1 = 0.4434
*AUC is  0.8202512069321625



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  PhysicalHealth
accuracy = 72.6683%
precision = 31.3013%
recall = 0.7726
f1 = 0.4455
*AUC is  0.8218875655141393



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  HardToClimbStairs
accuracy = 72.7373%
precision = 31.2850%
recall = 0.7674
f1 = 0.4445
*AUC is  0.8193180117395156



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  BiologicalSex
accuracy = 72.4101%
precision = 31.0345%
recall = 0.7701
f1 = 0.4424
*AUC is  0.819157052403585



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  AgeBracket
accuracy = 72.8694%
precision = 31.2507%
recall = 0.7575
f1 = 0.4425
*AUC is  0.8159578969872676



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  EducationBracket
accuracy = 72.6033%
precision = 31.2244%
recall = 0.7713
f1 = 0.4445
*AUC is  0.8214654138632893



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  IncomeBracket
accuracy = 72.4456%
precision = 31.0932%
recall = 0.7719
f1 = 0.4433
*AUC is  0.8187911856216223

After dropping  Zodiac
accuracy = 72.4732%
precision = 31.1807%
recall = 0.7760
f1 = 0.4449
*AUC is  0.8208049066017032



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


1. Build a logistic regression model. Doing so: What is the best predictor of diabetes and what is the AUC of this model?

I wrote a logistic regression function. I split data into training and testing sets and fitted the logistic regression with balanced class_weight. Finally, I computed evaluation metrics and the AUC. There are 22 columns including diabetes. I build 22 models, 1 with all predictors and 21 with 1 predictor dropped

I put balanced for class_weight so that the parameter adjusts the weights of the different classes. The evaluation metrics and AUC evaluate the performance of a model.

The data show that GeneralHealth is the best predictor, the reason is the AUC decreases the most when we drop GeneralHealth. The AUC decrease from 0.825 to 0.806(the lowest among the 21 models)

I believe the findings mean that the dropped column is important in predicting the target variable. Maybe it has the strongest correlation between the predictor and the target variable.


In [81]:
from sklearn.model_selection import train_test_split
from sklearn.svm import LinearSVC
from sklearn import metrics

def svm_classification(X, y, random_state, name):
    x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=random_state)
    model = LinearSVC(dual=False, class_weight='balanced', random_state=0).fit(x_train, y_train)
    y_pred = model.predict(x_test)
    accuracy = metrics.accuracy_score(y_test, y_pred)
    precision = metrics.precision_score(y_test, y_pred)
    recall = metrics.recall_score(y_test, y_pred)
    f1 = metrics.f1_score(y_test, y_pred)
    if name != '':
        print(f'After dropping {name}:')
    print(f'accuracy = {100 * accuracy:.4f}%')
    print(f'precision = {100 * precision:.4f}%')
    print(f'recall = {recall:.4f}')
    print(f'f1 = {f1:.4f}')
    y_prob = model.decision_function(x_test)
    fpr, tpr, thresholds = metrics.roc_curve(y_test, y_prob)
    auc_roc = metrics.auc(fpr, tpr)
    print(f'*AUC is {auc_roc:.4f}')
    print('')
    return model


In [82]:
model1 = svm_classification(scaled, y, 0, '')
for i in range(len(predictors)):
    temp = predictors[:]
    temp.pop(i)
    model2 = svm_classification(df[temp], y, 0,predictors[i])

accuracy = 72.7255%
precision = 31.4047%
recall = 0.7760
f1 = 0.4471
*AUC is 0.8249

After dropping HighBP:
accuracy = 72.6624%
precision = 31.1285%
recall = 0.7616
f1 = 0.4419
*AUC is 0.8176

After dropping HighChol:
accuracy = 72.4417%
precision = 31.0601%
recall = 0.7699
f1 = 0.4426
*AUC is 0.8190

After dropping BMI:
accuracy = 71.6651%
precision = 30.2845%
recall = 0.7631
f1 = 0.4336
*AUC is 0.8101

After dropping Smoker:
accuracy = 72.7531%
precision = 31.4398%
recall = 0.7767
f1 = 0.4476
*AUC is 0.8249

After dropping Stroke:
accuracy = 72.6979%
precision = 31.3717%
recall = 0.7755
f1 = 0.4467
*AUC is 0.8246

After dropping Myocardial:
accuracy = 72.4732%
precision = 31.2079%
recall = 0.7778
f1 = 0.4454
*AUC is 0.8242

After dropping PhysActivity:
accuracy = 72.7235%
precision = 31.4050%
recall = 0.7762
f1 = 0.4472
*AUC is 0.8248

After dropping Fruit:
accuracy = 72.7294%
precision = 31.4040%
recall = 0.7758
f1 = 0.4471
*AUC is 0.8249

After dropping Vegetables:
accuracy = 72.73

2. Build a SVM. Doing so: What is the best predictor of diabetes and what is the AUC of this model?

I wrote a svm_classification function. In the function, I split data into training and testing sets and fitted the LinearSVC with balanced class_weight and False for dual. Finally, I computed evaluation metrics and the AUC. There are 22 columns including diabetes. I build 22 models, 1 with all predictors and 21 with 1 predictor dropped

I put balanced for class_weight so that the parameter adjusts the weights of the different classes and False to dual to tell the algorithm to solve the primal optimization problem. The evaluation metrics and AUC evaluate the performance of a model.

Based on the data, it has been observed that GeneralHealth is the most significant predictor, as it leads to the highest decrease in AUC when dropped. The AUC drops from 0.8249 to 0.8096, which is the lowest among all 21 models. 

These findings suggest that GeneralHealth is a crucial factor in predicting the target variable in SVM model, possibly due to its strong correlation with the predictor variable.

In [83]:
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn import metrics
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
param_grid = {
    'criterion': ['entropy', 'gini'],
    'max_depth': [2, 5, 10, 20],
    'min_samples_split': [2, 5],
    'max_leaf_nodes': [5, 10, 20, 50]
}

dt = DecisionTreeClassifier()
grid_search = GridSearchCV(estimator=dt, param_grid=param_grid, cv=5, n_jobs=-1)
grid_search.fit(x_train, y_train)
print('Best hyperparameters:', grid_search.best_params_)
print('Best cross-validation score:', grid_search.best_score_)


Best hyperparameters: {'criterion': 'gini', 'max_depth': 10, 'max_leaf_nodes': 20, 'min_samples_split': 2}
Best cross-validation score: 0.8654308542398607


In [84]:


def decision_tree_classification(X, y, random_state, name):
    x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
    model = DecisionTreeClassifier(criterion='gini', class_weight = 'balanced', max_features=1.0, max_leaf_nodes=20).fit(x_train, y_train)
    y_pred = model.predict(x_test)
    accuracy = metrics.accuracy_score(y_test, y_pred)
    precision = metrics.precision_score(y_test, y_pred)
    recall = metrics.recall_score(y_test, y_pred)
    f1 = metrics.f1_score(y_test, y_pred)
    if name != '':
        print(f'After dropping {name}:')
    print(f'accuracy = {100 * accuracy:.4f}%')
    print(f'precision = {100 * precision:.4f}%')
    print(f'recall = {recall:.4f}')
    print(f'f1 = {f1:.4f}')
    # Compute AUC-ROC
    y_prob = model.predict_proba(x_test)[:, 1]
    fpr, tpr, thresholds = metrics.roc_curve(y_test, y_prob)
    auc_roc = metrics.auc(fpr, tpr)
    print(f'*AUC is {auc_roc:.4f}')
    print('')
    return model


In [85]:

model1 = decision_tree_classification(scaled, y, 0, '')
for i in range(len(predictors)):
    temp = predictors[:]
    temp.pop(i)
    model2 = decision_tree_classification(df[temp], y, 0,predictors[i])

accuracy = 72.4377%
precision = 30.8011%
recall = 0.7534
f1 = 0.4373
*AUC is 0.8045

After dropping HighBP:
accuracy = 65.9768%
precision = 27.0347%
recall = 0.8204
f1 = 0.4067
*AUC is 0.7990

After dropping HighChol:
accuracy = 72.4377%
precision = 30.8011%
recall = 0.7534
f1 = 0.4373
*AUC is 0.8043

After dropping BMI:
accuracy = 65.8034%
precision = 26.8400%
recall = 0.8147
f1 = 0.4038
*AUC is 0.7985

After dropping Smoker:
accuracy = 72.4377%
precision = 30.8011%
recall = 0.7534
f1 = 0.4373
*AUC is 0.8045

After dropping Stroke:
accuracy = 72.4377%
precision = 30.8011%
recall = 0.7534
f1 = 0.4373
*AUC is 0.8045

After dropping Myocardial:
accuracy = 72.4377%
precision = 30.8011%
recall = 0.7534
f1 = 0.4373
*AUC is 0.8045

After dropping PhysActivity:
accuracy = 72.4377%
precision = 30.8011%
recall = 0.7534
f1 = 0.4373
*AUC is 0.8045

After dropping Fruit:
accuracy = 72.4377%
precision = 30.8011%
recall = 0.7534
f1 = 0.4373
*AUC is 0.8045

After dropping Vegetables:
accuracy = 72.43

3. Use a single, individual decision tree. Doing so: What is the best predictor of diabetes and what is the AUC of this model?

I utilized GridSearchCV to find the best parameters for the model to fit the data. After collecting the parameters, I wrote a decision_tree_classification function. In the function, I split data into training and testing sets and fitted the DecisionTreeClassifier with balanced class_weight, gini criterion, max_features=1.0 and max_leaf_nodes=20. Finally, I computed evaluation metrics and the AUC. There are 22 columns including diabetes. I build 22 models, 1 with all predictors and 21 with 1 predictor dropped

I put balanced for class_weight so that the parameter adjusts the weights of the different classes and False to dual to tell the algorithm to solve the primal optimization problem. Set a number for max_features and max_leaf_nodes controls the complexity of the decision tree and can help prevent overfitting. The evaluation metrics and AUC evaluate the performance of a model.

According to the data, the most notable predictor is GeneralHealth, as removing this predictor leads to the largest reduction in AUC among all 21 models tested. Specifically, the AUC value decreases from 0.8045 to 0.7874, which is the lowest observed in the model.

These findings indicate that GeneralHealth plays a crucial role in predicting the target variable in the decision tree model, likely due to its high correlation with the predictor variable.

In [86]:
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn import metrics
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
param_grid = {
    'criterion': ['entropy', 'gini'],
    'n_estimators': [10, 50, 100],
    'max_samples': [ 0.01, 1.0],
    'max_features': ['sqrt', 'log2'],
    'bootstrap':[True, False],
    'max_depth': [2, 5, 10],
    'min_samples_split': [2, 10, 50],
    'min_samples_leaf': [1, 4]
}
dt = RandomForestClassifier()
grid_search = GridSearchCV(estimator=dt, param_grid=param_grid, cv=5, n_jobs=-1)
grid_search.fit(x_train, y_train)
print('Best hyperparameters:', grid_search.best_params_)
print('Best cross-validation score:', grid_search.best_score_)


/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:378: FitFailedWarning: 
2160 fits failed out of a total of 4320.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
2160 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_forest.py", line 397, in fit
    raise ValueError(
ValueError: `max_sample` cannot be set if `bootstrap=False`. Either switch to `bootstrap=True` or set `max_sample=None`.

  warnin

Best hyperparameters: {'bootstrap': True, 'criterion': 'gini', 'max_depth': 10, 'max_features': 'sqrt', 'max_samples': 1.0, 'min_samples_leaf': 1, 'min_samples_split': 10, 'n_estimators': 100}
Best cross-validation score: 0.8660911357825739


In [87]:


def random_forest_classification(X, y, random_state, name):
    x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=random_state)
    model = RandomForestClassifier(n_estimators=50, max_samples=1.0,bootstrap=True, criterion='gini', max_depth=10, max_features = 'sqrt', min_samples_leaf = 1,min_samples_split = 2).fit(x_train, y_train)#max_leaf_nodes min_sample_leaf
    y_pred = model.predict(x_test)
    accuracy = metrics.accuracy_score(y_test, y_pred)
    precision = metrics.precision_score(y_test, y_pred)
    recall = metrics.recall_score(y_test, y_pred)
    f1 = metrics.f1_score(y_test, y_pred)
    if name != '':
        print(f'After dropping {name}:')
    print(f'accuracy = {100 * accuracy:.4f}%')
    print(f'precision = {100 * precision:.4f}%')
    print(f'recall = {recall:.4f}')
    print(f'f1 = {f1:.4f}')
    y_prob = model.predict_proba(x_test)[:, 1]
    fpr, tpr, thresholds = metrics.roc_curve(y_test, y_prob)
    auc_roc = metrics.auc(fpr, tpr)
    print(f'*AUC is {auc_roc:.4f}')
    print('')
    return model


In [88]:
model1 = random_forest_classification(scaled, y, 0, '')
for i in range(len(predictors)):
    temp = predictors[:]
    temp.pop(i)
    model2 = random_forest_classification(df[temp], y, 0,predictors[i])

accuracy = 86.3273%
precision = 63.9796%
recall = 0.0870
f1 = 0.1531
*AUC is 0.8288

After dropping HighBP:
accuracy = 86.2800%
precision = 64.0449%
recall = 0.0790
f1 = 0.1407
*AUC is 0.8229

After dropping HighChol:
accuracy = 86.2228%
precision = 64.6358%
recall = 0.0677
f1 = 0.1225
*AUC is 0.8233

After dropping BMI:
accuracy = 85.9646%
precision = 62.5698%
recall = 0.0311
f1 = 0.0592
*AUC is 0.8152

After dropping Smoker:
accuracy = 86.3371%
precision = 63.9083%
recall = 0.0889
f1 = 0.1561
*AUC is 0.8285

After dropping Stroke:
accuracy = 86.3726%
precision = 63.8140%
recall = 0.0951
f1 = 0.1656
*AUC is 0.8287

After dropping Myocardial:
accuracy = 86.3016%
precision = 63.6364%
recall = 0.0845
f1 = 0.1491
*AUC is 0.8289

After dropping PhysActivity:
accuracy = 86.3746%
precision = 64.1098%
recall = 0.0939
f1 = 0.1638
*AUC is 0.8288

After dropping Fruit:
accuracy = 86.3312%
precision = 63.3981%
recall = 0.0906
f1 = 0.1585
*AUC is 0.8292

After dropping Vegetables:
accuracy = 86.40

4. Build a random forest model. Doing so: What is the best predictor of diabetes and what is the AUC of this model?

I also utilized GridSearchCV to find the best parameters for the model to fit the data. After collecting the parameters, I wrote a random_forest_classification function. In the function, I split data into training and testing sets and fitted the RandomForestClassifier with balanced class_weight, gini criterion, True to bootstrap, max_leaf_nodes=20, n_estimators=50, max_samples=1.0, max_depth=10, max_features = 'sqrt', min_samples_leaf = 1 and min_samples_split = 2. Finally, I computed evaluation metrics and the AUC. There are 22 columns including diabetes. I build 22 models, 1 with all predictors and 21 with 1 predictor dropped

Balanced class_weight was chosen to adjust the weights of different classes, while the number of max_features, min_samples_split, min_samples_leaf, max_leaf_nodes, and n_estimators were set to control decision tree complexity and prevent overfitting. Evaluation metrics and AUC were utilized to assess the model's performance.

The results showed that GeneralHealth was the most significant predictor, as removing it led to the largest reduction in AUC among all 21 models tested. Specifically, the AUC value decreases from 0.8289 to 0.8143, which is the lowest observed in the model.

This indicates that GeneralHealth plays a crucial role in predicting the target variable, likely due to its high correlation with the predictor variable.


In [89]:
from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier

base_estimator = DecisionTreeClassifier(max_depth=1)
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
param_grid = {
    'n_estimators': [50, 100, 200],
    'learning_rate': [0.01, 0.1, 1],
    'algorithm': ['SAMME', 'SAMME.R']
}
ada = AdaBoostClassifier(estimator=base_estimator)
grid_search = GridSearchCV(estimator=ada, param_grid=param_grid, cv=5, scoring='roc_auc')
grid_search.fit(X, y)
print(f"Best parameters: {grid_search.best_params_}")
print(f"Best score: {grid_search.best_score_:.4f}")


Best parameters: {'algorithm': 'SAMME.R', 'learning_rate': 1, 'n_estimators': 200}
Best score: 0.8256


In [90]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import AdaBoostClassifier
from sklearn import metrics
from sklearn.tree import DecisionTreeClassifier

def adaboost_classification(X, y, random_state, name):
    x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=random_state)
    bdt = AdaBoostClassifier(
        DecisionTreeClassifier(criterion='gini', class_weight = 'balanced', max_features=1.0, max_leaf_nodes=20), algorithm="SAMME.R", n_estimators=200, learning_rate=0.1
    )
    model = bdt.fit(x_train, y_train)
    y_pred = model.predict(x_test)
    accuracy = metrics.accuracy_score(y_test, y_pred)
    precision = metrics.precision_score(y_test, y_pred)
    recall = metrics.recall_score(y_test, y_pred)
    f1 = metrics.f1_score(y_test, y_pred)
    if name != '':
        print(f'After dropping {name}:')
    print(f'accuracy = {100 * accuracy:.4f}%')
    print(f'precision = {100 * precision:.4f}%')
    print(f'recall = {recall:.4f}')
    print(f'f1 = {f1:.4f}')
    y_prob = model.predict_proba(x_test)[:, 1]
    fpr, tpr, thresholds = metrics.roc_curve(y_test, y_prob)
    auc_roc = metrics.auc(fpr, tpr)
    print(f'*AUC is {auc_roc:.4f}')
    print('')
    return model


In [91]:

model1 = adaboost_classification(scaled, y, 0, '')
for i in range(len(predictors)):
    temp = predictors[:]
    temp.pop(i)
    model2 = adaboost_classification(df[temp], y, 0,predictors[i])

accuracy = 71.6809%
precision = 30.7928%
recall = 0.7956
f1 = 0.4440
*AUC is 0.8231

After dropping HighBP:
accuracy = 70.9910%
precision = 30.2208%
recall = 0.7953
f1 = 0.4380
*AUC is 0.8197

After dropping HighChol:
accuracy = 71.2295%
precision = 30.3960%
recall = 0.7941
f1 = 0.4396
*AUC is 0.8197

After dropping BMI:
accuracy = 70.0666%
precision = 29.4624%
recall = 0.7934
f1 = 0.4297
*AUC is 0.8125

After dropping Smoker:
accuracy = 71.7577%
precision = 30.8574%
recall = 0.7956
f1 = 0.4447
*AUC is 0.8248

After dropping Stroke:
accuracy = 71.7420%
precision = 30.8379%
recall = 0.7952
f1 = 0.4444
*AUC is 0.8234

After dropping Myocardial:
accuracy = 71.5725%
precision = 30.7207%
recall = 0.7968
f1 = 0.4434
*AUC is 0.8217

After dropping PhysActivity:
accuracy = 71.5843%
precision = 30.7037%
recall = 0.7950
f1 = 0.4430
*AUC is 0.8229

After dropping Fruit:
accuracy = 71.5725%
precision = 30.7434%
recall = 0.7984
f1 = 0.4439
*AUC is 0.8248

After dropping Vegetables:
accuracy = 71.69

5. Build a model using adaBoost. Doing so: What is the best predictor of diabetes and what is the AUC of this model?

To improve the performance of the AdaBoostClassifier, I used GridSearchCV to find the optimal hyperparameters. I wrote an adaboost_classification function that split the data into training and testing sets and fitted the AdaBoostClassifier using balanced class_weight, algorithm="SAMME.R", n_estimators=200, and learning_rate=0.1. After evaluating the model using evaluation metrics and AUC, I built 22 models, including one with all predictors and 21 with one predictor removed.

By using balanced class_weight, the weights of different classes were adjusted, while learning_rate controlled the contribution of each estimator to the final prediction. Algorithm was used to update the weights of the misclassified samples in each iteration. Evaluation metrics and AUC were used to assess the performance of the model.

The analysis showed that GeneralHealth was the most important predictor, as removing it led to the largest decrease in AUC among all the models. Specifically, the AUC decreased from 0.8231 to 0.8112, which was the lowest observed in the model. 

This suggests that GeneralHealth plays a critical role in predicting the target variable due to its strong correlation with the predictor variable.

a) Which of these 5 models is the best to predict diabetes in this dataset?

After evaluating the AUC values of 5 models, it was observed that the AUC values for all models, except for the decision tree model, ranged from 0.82 to 0.83. The decision tree model had an AUC value of approximately 80. I guess the reason is I did not find the best parameter for decision tree model

I selected AUC as the primary evaluation metric instead of accuracy based on my professor's example in class. The example demonstrated that in certain cases, such as identifying phone scams, it may be better to prioritize minimizing false negatives even if it means accepting a high rate of false positives. AUC is a robust metric that can accurately reflect the performance of different models in such situations.

In [92]:
pd.set_option('display.max_columns', None)
df.corr()

Diabetes    HighBP  HighChol       BMI    Smoker  \
Diabetes               1.000000  0.263129  0.200276  0.216843  0.060789   
HighBP                 0.263129  1.000000  0.298199  0.213748  0.096991   
HighChol               0.200276  0.298199  1.000000  0.106722  0.091299   
BMI                    0.216843  0.213748  0.106722  1.000000  0.013804   
Smoker                 0.060789  0.096991  0.091299  0.013804  1.000000   
Stroke                 0.105816  0.129575  0.092620  0.020153  0.061173   
Myocardial             0.177282  0.209361  0.180765  0.052904  0.114441   
PhysActivity          -0.118133 -0.125267 -0.078046 -0.147294 -0.087401   
Fruit                 -0.040779 -0.040555 -0.040859 -0.087518 -0.077666   
Vegetables            -0.056584 -0.061266 -0.039874 -0.062275 -0.030678   
HeavyDrinker          -0.057056 -0.003972 -0.011543 -0.048736  0.101619   
HasHealthcare          0.016255  0.038425  0.042230 -0.018471 -0.023251   
NotAbleToAffordDoctor  0.031433  0.017358  0.013310  0.058206  0.048946   
GeneralHealth          0.293569  0.300530  0.208426  0.239185  0.163143   
MentalHealth           0.069315  0.056456  0.062069  0.085310  0.092196   
PhysicalHealth         0.171337  0.161212  0.121751  0.121141  0.116460   
HardToClimbStairs      0.218344  0.223618  0.144672  0.197078  0.122463   
BiologicalSex          0.031430  0.052207  0.031205  0.042950  0.093662   
AgeBracket             0.177442  0.344452  0.272318 -0.036618  0.120641   
EducationBracket      -0.124456 -0.141358 -0.070802 -0.103932 -0.161955   
IncomeBracket         -0.163919 -0.171235 -0.085459 -0.100069 -0.123937   
Zodiac                -0.000197 -0.002629  0.001052 -0.001932  0.000975   

                         Stroke  Myocardial  PhysActivity     Fruit  \
Diabetes               0.105816    0.177282     -0.118133 -0.040779   
HighBP                 0.129575    0.209361     -0.125267 -0.040555   
HighChol               0.092620    0.180765     -0.078046 -0.040859   
BMI                    0.020153    0.052904     -0.147294 -0.087518   
Smoker                 0.061173    0.114441     -0.087401 -0.077666   
Stroke                 1.000000    0.203002     -0.069151 -0.013389   
Myocardial             0.203002    1.000000     -0.087299 -0.019790   
PhysActivity          -0.069151   -0.087299      1.000000  0.142756   
Fruit                 -0.013389   -0.019790      0.142756  1.000000   
Vegetables            -0.041124   -0.039167      0.153150  0.254342   
HeavyDrinker          -0.016950   -0.028991      0.012392 -0.035288   
HasHealthcare          0.008776    0.018734      0.035505  0.031544   
NotAbleToAffordDoctor  0.034804    0.031000     -0.061638 -0.044243   
GeneralHealth          0.177942    0.258383     -0.266186 -0.103854   
MentalHealth           0.070172    0.064621     -0.125587 -0.068217   
PhysicalHealth         0.148944    0.181698     -0.219230 -0.044633   
HardToClimbStairs      0.176567    0.212709     -0.253174 -0.048352   
BiologicalSex          0.002978    0.086096      0.032482 -0.091175   
AgeBracket             0.126974    0.221618     -0.092511  0.064547   
EducationBracket      -0.076009   -0.099600      0.199658  0.110187   
IncomeBracket         -0.128599   -0.141011      0.198539  0.079929   
Zodiac                -0.001933   -0.001232     -0.000842 -0.002255   

                       Vegetables  HeavyDrinker  HasHealthcare  \
Diabetes                -0.056584     -0.057056       0.016255   
HighBP                  -0.061266     -0.003972       0.038425   
HighChol                -0.039874     -0.011543       0.042230   
BMI                     -0.062275     -0.048736      -0.018471   
Smoker                  -0.030678      0.101619      -0.023251   
Stroke                  -0.041124     -0.016950       0.008776   
Myocardial              -0.039167     -0.028991       0.018734   
PhysActivity             0.153150      0.012392       0.035505   
Fruit                    0.254342     -0.035288       0.031544   
Veg

In [93]:
corr_matrix = df.corr()
mask = (abs(corr_matrix) >= 0.4) & (corr_matrix != 1)

filtered_corr = corr_matrix.where(mask).stack()

print(filtered_corr)

GeneralHealth      PhysicalHealth       0.524364
                   HardToClimbStairs    0.456920
PhysicalHealth     GeneralHealth        0.524364
                   HardToClimbStairs    0.478417
HardToClimbStairs  GeneralHealth        0.456920
                   PhysicalHealth       0.478417
EducationBracket   IncomeBracket        0.449106
IncomeBracket      EducationBracket     0.449106
dtype: float64


In [94]:
df[['BMI', 'GeneralHealth']]

BMI  GeneralHealth
0        40              5
1        25              3
2        28              5
3        27              2
4        24              2
...     ...            ...
253675   45              3
253676   18              4
253677   28              1
253678   23              3
253679   25              2

[253680 rows x 2 columns]

In [95]:
df.head()

Diabetes  HighBP  HighChol  BMI  Smoker  Stroke  Myocardial  PhysActivity  \
0         0       1         1   40       1       0           0             0   
1         0       0         0   25       1       0           0             1   
2         0       1         1   28       0       0           0             0   
3         0       1         0   27       0       0           0             1   
4         0       1         1   24       0       0           0             1   

   Fruit  Vegetables  HeavyDrinker  HasHealthcare  NotAbleToAffordDoctor  \
0      0           1             0              1                      0   
1      0           0             0              0                      1   
2      1           0             0              1                      1   
3      1           1             0              1                      0   
4      1           1             0              1                      0   

   GeneralHealth  MentalHealth  PhysicalHealth  HardToClimbStairs  \
0              5            18              15                  1   
1              3             0               0                  0   
2              5            30              30                  1   
3              2             0               0                  0   
4              2             3               0                  0   

   BiologicalSex  AgeBracket  EducationBracket  IncomeBracket  Zodiac  
0              1           9                 4              3      10  
1              1           7                 6              1      11  
2              1           9                 4              8       2  
3              1          11                 3              6      11  
4              1          11                 5              4       8

In [96]:
columns_to_drop = ['IncomeBracket', 'HardToClimbStairs', 'GeneralHealth', 'PhysicalHealth']
df1 = df.drop(columns=columns_to_drop)


In [97]:
predictors = list(df1.columns)
predictors.remove('Diabetes')
X = df1[predictors]
y = df1['Diabetes']
scaled = StandardScaler().fit_transform(X)
scaled = pd.DataFrame(scaled, columns=X.columns)
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [98]:

model1 = logistic_regersion(scaled, y, 0, '')
for i in range(len(predictors)):
    temp = predictors[:]
    temp.pop(i)
    model2 = logistic_regersion(df1[temp], y, 0,predictors[i])

accuracy = 71.2610%
precision = 29.8016%
recall = 0.7540
f1 = 0.4272
*AUC is  0.8044369414237343



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  HighBP
accuracy = 70.5061%
precision = 28.8481%
recall = 0.7332
f1 = 0.4140
*AUC is  0.7904152894880827



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  HighChol
accuracy = 70.6619%
precision = 29.2639%
recall = 0.7509
f1 = 0.4212
*AUC is  0.7963659180789195



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  BMI
accuracy = 69.6626%
precision = 28.2362%
recall = 0.7360
f1 = 0.4081
*AUC is  0.7789721013854082



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  Smoker
accuracy = 70.9871%
precision = 29.6029%
recall = 0.7557
f1 = 0.4254
*AUC is  0.8036182292497637



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  Stroke
accuracy = 70.8629%
precision = 29.4596%
recall = 0.7530
f1 = 0.4235
*AUC is  0.8018382551196791



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  Myocardial
accuracy = 70.6284%
precision = 29.2684%
recall = 0.7529
f1 = 0.4215
*AUC is  0.80134840263746



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  PhysActivity
accuracy = 71.0028%
precision = 29.6268%
recall = 0.7563
f1 = 0.4258
*AUC is  0.803236023909161



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  Fruit
accuracy = 70.9634%
precision = 29.5503%
recall = 0.7536
f1 = 0.4245
*AUC is  0.8024551504096071



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  Vegetables
accuracy = 71.1684%
precision = 29.6850%
recall = 0.7515
f1 = 0.4256
*AUC is  0.803400468889232



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  HeavyDrinker
accuracy = 70.6954%
precision = 29.3645%
recall = 0.7555
f1 = 0.4229
*AUC is  0.8016342987383418



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  HasHealthcare
accuracy = 71.1014%
precision = 29.6654%
recall = 0.7537
f1 = 0.4257
*AUC is  0.8034127801588469



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  NotAbleToAffordDoctor
accuracy = 70.9003%
precision = 29.4789%
recall = 0.7523
f1 = 0.4236
*AUC is  0.8030498111770175



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  MentalHealth
accuracy = 71.1250%
precision = 29.6671%
recall = 0.7526
f1 = 0.4256
*AUC is  0.8034930598961268



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  BiologicalSex
accuracy = 71.2315%
precision = 29.7327%
recall = 0.7512
f1 = 0.4260
*AUC is  0.8038985400360401



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  AgeBracket
accuracy = 71.0265%
precision = 29.2491%
recall = 0.7319
f1 = 0.4180
*AUC is  0.7935520537859743



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


After dropping  EducationBracket
accuracy = 70.6402%
precision = 29.3158%
recall = 0.7552
f1 = 0.4224
*AUC is  0.7997941207793218

After dropping  Zodiac
accuracy = 70.6500%
precision = 29.3903%
recall = 0.7594
f1 = 0.4238
*AUC is  0.8026065274102978



/Users/yanggezheng/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [99]:

model1 = svm_classification(scaled, y, 0, '')
for i in range(len(predictors)):
    temp = predictors[:]
    temp.pop(i)
    model2 = svm_classification(df1[temp], y, 0,predictors[i])

accuracy = 70.7742%
precision = 29.4723%
recall = 0.7583
f1 = 0.4245
*AUC is 0.8038

After dropping HighBP:
accuracy = 70.3958%
precision = 28.8340%
recall = 0.7376
f1 = 0.4146
*AUC is 0.7910

After dropping HighChol:
accuracy = 69.9996%
precision = 28.8632%
recall = 0.7584
f1 = 0.4181
*AUC is 0.7958

After dropping BMI:
accuracy = 69.2152%
precision = 27.9872%
recall = 0.7412
f1 = 0.4063
*AUC is 0.7786

After dropping Smoker:
accuracy = 70.7624%
precision = 29.4406%
recall = 0.7569
f1 = 0.4239
*AUC is 0.8037

After dropping Stroke:
accuracy = 70.6579%
precision = 29.3567%
recall = 0.7569
f1 = 0.4231
*AUC is 0.8028

After dropping Myocardial:
accuracy = 70.0942%
precision = 28.9454%
recall = 0.7590
f1 = 0.4191
*AUC is 0.8008

After dropping PhysActivity:
accuracy = 70.7072%
precision = 29.4051%
recall = 0.7575
f1 = 0.4236
*AUC is 0.8031

After dropping Fruit:
accuracy = 70.7998%
precision = 29.4819%
recall = 0.7576
f1 = 0.4245
*AUC is 0.8038

After dropping Vegetables:
accuracy = 70.71

In [100]:

model1 = decision_tree_classification(scaled, y, 0, '')
for i in range(len(predictors)):
    temp = predictors[:]
    temp.pop(i)
    model2 = decision_tree_classification(df1[temp], y, 0,predictors[i])

accuracy = 65.6457%
precision = 26.5802%
recall = 0.8042
f1 = 0.3995
*AUC is 0.7820

After dropping HighBP:
accuracy = 68.5963%
precision = 27.1316%
recall = 0.7175
f1 = 0.3937
*AUC is 0.7705

After dropping HighChol:
accuracy = 62.1570%
precision = 25.0634%
recall = 0.8355
f1 = 0.3856
*AUC is 0.7772

After dropping BMI:
accuracy = 67.5753%
precision = 27.0172%
recall = 0.7532
f1 = 0.3977
*AUC is 0.7689

After dropping Smoker:
accuracy = 65.6457%
precision = 26.5802%
recall = 0.8042
f1 = 0.3995
*AUC is 0.7820

After dropping Stroke:
accuracy = 65.6457%
precision = 26.5802%
recall = 0.8042
f1 = 0.3995
*AUC is 0.7820

After dropping Myocardial:
accuracy = 66.6470%
precision = 26.9148%
recall = 0.7851
f1 = 0.4009
*AUC is 0.7815

After dropping PhysActivity:
accuracy = 65.6457%
precision = 26.5802%
recall = 0.8042
f1 = 0.3995
*AUC is 0.7820

After dropping Fruit:
accuracy = 65.6457%
precision = 26.5802%
recall = 0.8042
f1 = 0.3995
*AUC is 0.7820

After dropping Vegetables:
accuracy = 65.64

In [101]:

model1 = random_forest_classification(scaled, y, 0, '')
for i in range(len(predictors)):
    temp = predictors[:]
    temp.pop(i)
    model2 = random_forest_classification(df1[temp], y, 0,predictors[i])

accuracy = 86.1361%
precision = 63.3484%
recall = 0.0582
f1 = 0.1067
*AUC is 0.8085

After dropping HighBP:
accuracy = 86.0769%
precision = 63.9469%
recall = 0.0467
f1 = 0.0871
*AUC is 0.7971

After dropping HighChol:
accuracy = 86.1105%
precision = 64.6953%
recall = 0.0501
f1 = 0.0929
*AUC is 0.8012

After dropping BMI:
accuracy = 85.8207%
precision = 55.2147%
recall = 0.0125
f1 = 0.0244
*AUC is 0.7853

After dropping Smoker:
accuracy = 86.1499%
precision = 63.0682%
recall = 0.0616
f1 = 0.1122
*AUC is 0.8078

After dropping Stroke:
accuracy = 86.1243%
precision = 61.6327%
recall = 0.0628
f1 = 0.1140
*AUC is 0.8081

After dropping Myocardial:
accuracy = 86.0296%
precision = 63.1130%
recall = 0.0410
f1 = 0.0771
*AUC is 0.8057

After dropping PhysActivity:
accuracy = 86.1755%
precision = 65.1772%
recall = 0.0587
f1 = 0.1076
*AUC is 0.8074

After dropping Fruit:
accuracy = 86.1913%
precision = 64.7482%
recall = 0.0624
f1 = 0.1138
*AUC is 0.8085

After dropping Vegetables:
accuracy = 86.14

In [102]:

model1 = adaboost_classification(scaled, y, 0, '')
for i in range(len(predictors)):
    temp = predictors[:]
    temp.pop(i)
    model2 = adaboost_classification(df1[temp], y, 0,predictors[i])

accuracy = 70.1770%
precision = 29.2844%
recall = 0.7763
f1 = 0.4253
*AUC is 0.8054

After dropping HighBP:
accuracy = 69.2171%
precision = 28.3045%
recall = 0.7605
f1 = 0.4125
*AUC is 0.7912

After dropping HighChol:
accuracy = 69.4162%
precision = 28.6632%
recall = 0.7737
f1 = 0.4183
*AUC is 0.7949

After dropping BMI:
accuracy = 68.0247%
precision = 27.6377%
recall = 0.7723
f1 = 0.4071
*AUC is 0.7811

After dropping Smoker:
accuracy = 70.1533%
precision = 29.1373%
recall = 0.7681
f1 = 0.4225
*AUC is 0.8000

After dropping Stroke:
accuracy = 70.2006%
precision = 29.2811%
recall = 0.7749
f1 = 0.4250
*AUC is 0.8044

After dropping Myocardial:
accuracy = 69.7158%
precision = 28.9389%
recall = 0.7769
f1 = 0.4217
*AUC is 0.8013

After dropping PhysActivity:
accuracy = 70.2933%
precision = 29.2815%
recall = 0.7704
f1 = 0.4243
*AUC is 0.8036

After dropping Fruit:
accuracy = 70.2361%
precision = 29.2848%
recall = 0.7734
f1 = 0.4248
*AUC is 0.8025

After dropping Vegetables:
accuracy = 70.17

b) Tell us something interesting about this dataset that is not already covered by the questions above and that is not obvious.

After examining the correlation between each column, it was found that some columns, such as PhysicalHealth, GeneralHealth, and HardToClimbStairs, are highly correlated, which may result in double counting of factors. 

Additionally, some of the columns contain self-assessment data, which are known to be less reliable due to various biases and limitations compared to objective assessments or observations. For instance, a person in the dataframe who has high blood pressure, high cholesterol, smokes, lacks physical activity, does not eat fruit, and has a high BMI, but still rates their general health as "very healthy" (score of 5).

After removing self-assessed and correlated data, the analysis was repeated and it was found that BMI was the most accurate predictor of diabetes.

 A possible reason for this is that BMI has a wider range of values compared to other variables which are mostly binary. For example, Age bracket has only six options. Providing more options can lead to more precise and accurate feedback by allowing respondents to better differentiate between their preferences and experiences, resulting in more detailed and nuanced feedback. This can also help capture a wider range of responses, increasing the diversity and richness of the feedback.